### PR TITLE
Feature/support multiple actions

### PR DESCRIPTION
**Pull request by Kureev**
At the current moment, we have 1 to 1 plugin-command relation. But sometimes it's not enough: for some plugins you want to expose multiple actions (like for `rnpm-plugin-link` you want to expose `link` and `unlink` command). This PR will allow to do this.

### CODE REVIEW

**Comment by grabbou** [comment]
Can we get a test suite for that to make sure the existing behaviour does not break?

**Comment by Kureev** [comment]
Sure, will add it today

**Comment by Kureev** [comment]
hey @grabbou, I've added 2 new tests and made some changes (see commit message).
If you prefer old naming, feel free to roll it back (but I'd keep it like it is now). 

Ping me if you need any explanations.

File: bin/cli [hunk 1]
```diff
@@ -14,17 +14,24 @@ log.heading = 'rnpm';
...
+    .description(command.description)
     .action(function runAction() {
-      action.func(config, arguments);
-    })
-);
+      command.func(config, arguments);
+    });
+
+commands.forEach((command) => {
+  if (isArray(command)) {
```
Comment (grabbou) [inline bin/cli]: I'd prefer something flatting that before passing here.

File: test/fixtures/files/plugins.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+module.exports = [{
+  func: () => {},
+  description: 'Test action #1',
+  name: 'test',
+}, {
+  func: () => {},
+  description: 'Test action #2',
+  name: 'test2',
+}];
```
Comment (grabbou) [inline test/fixtures/files/plugins.js]: Can we add one test-spec for single non-array object being exported? Just like `rnpm-link` does for now.

File: test/getCommands.spec.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+const multiPlugin = require(path.join(filesPath, 'plugins'));
+
+describe('getCommands', () => {
+
+  before(() => {
+    mockRequire(path.join(__dirname, '..'), {});
+    mockRequire(
+      path.join(__dirname, '..', 'package.json'),
+      require(path.join(filesPath, 'package'))
+    );
```
Comment (grabbou) [inline test/getCommands.spec.js]: It's okey, but can't we just use `mock-fs` in that case? Because it's our test_suite, `__dirname/.../` equals `process.cwd()` anyway. That way, we don't have `mock-require` dependency at all.

Something like:
```js
mockFs({
   'package.json': require(path.join(filesPath, 'package'))
});
```
should be enough.

File: test/getCommands.spec.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+const expect = require('chai').expect;
+const getCommands = require('../src/getCommands');
+const findPlugins = require('../src/findPlugins');
+const mockRequire = require('mock-require');
+const mockFs = require('mock-fs');
+
+const commands = getCommands();
+const filesPath = path.join(__dirname, 'fixtures', 'files');
+const testPlugin = findPlugins(filesPath);
+const testPluginPath = path.join(__dirname, '..', 'node_modules', 'rnpm-plugin-test');
```
Comment (grabbou) [inline test/getCommands.spec.js]: There are bunch of unused vars here. Check that out.

File: test/getCommands.spec.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
+const path = require('path');
+const Module = require('module');
+const expect = require('chai').expect;
+const getCommands = require('../src/getCommands');
+const findPlugins = require('../src/findPlugins');
+const mockRequire = require('mock-require');
+const mockFs = require('mock-fs');
+
+const commands = getCommands();
```
Comment (grabbou) [inline test/getCommands.spec.js]: Can we call `getCommands()` in every suite separately instead of having global reference?

File: test/getCommands.spec.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    );
+  });
+
+  it('list of the commands should be an array', () => {
+    expect(commands).to.be.an('array');
+  });
+
+  it('should contain some pre-defined commands by default', () => {
+    expect(commands).to.be.not.empty;
+  });
```
Comment (grabbou) [inline test/getCommands.spec.js]: Looks like these cases are the same.

File: test/getCommands.spec.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  it('should return a single command (plugin export one command)', () => {
+    mockRequire(testPluginPath, singlePlugin);
+    expect(getCommands().length).to.be.equal(1);
+    mockRequire.stop();
+  });
+
+  it('should return a single command (plugin export one command)', () => {
+    mockRequire(testPluginPath, multiPlugin);
+    expect(getCommands().length).to.be.equal(2);
+    mockRequire.stop();
```
Comment (grabbou) [inline test/getCommands.spec.js]: Isn't it better to call that in `afterEach` ?

File: test/getCommands.spec.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  });
+
+  after(() => {
+    mockFs.restore();
```
Comment (grabbou) [inline test/getCommands.spec.js]: There's no `mockFs` to restore.

File: bin/cli (hunk 1 above)
Comment (Kureev) [inline bin/cli]: sure, I'll add `flatten` call here

File: test/fixtures/files/plugins.js (hunk 2 above)
Comment (Kureev) [inline test/fixtures/files/plugins.js]: I have a `plugin` for single and `plugins` for multi

File: test/getCommands.spec.js (hunk 3 above)
Comment (Kureev) [inline test/getCommands.spec.js]: In theory we can, but it leads us away from testing code as it is. If we'll change this behavior in the future (for some reason), this test wouldn't reflect this changes. So, I'd prefer to keep this one in.

File: test/fixtures/dependencies.js [hunk 9]
```diff
@@ -2,7 +2,7 @@ const fs = require('fs');
 const path = require('path');
 const android = require('./android');
 
-const pjson = fs.readFileSync(path.join(__dirname, './files/dependency.json'));
+const pjson = fs.readFileSync(path.join(__dirname, 'files', 'package.json'));
```
Comment (grabbou) [inline test/fixtures/dependencies.js]: As I said - I'd better keep that package.json related to `dependencies` w/o any extra plugins included. What I liked in my PR is that we were defining these plugins on per suite basis with simple `mock` which was way more explicit and allowed us to test out more cases (duplicates, both dev & production deps, unique names exported). But just saying :package:

File: test/getCommands.spec.js (hunk 3 above)
Comment (grabbou) [inline test/getCommands.spec.js]: That comment was invalid anyway since I didn't realise mock-fs does not work well with mock-fs.

File: test/fixtures/dependencies.js (hunk 9 above)
Comment (Kureev) [inline test/fixtures/dependencies.js]: I'd like not to rely on dependencies we have now. If we need to include something - mock it. Trust me, that will help us keep a code base consistent.

Comment (grabbou) [inline test/fixtures/dependencies.js]: But you are relying on the package.json that was made for dependency tests, not for plugin system. And that's bad stuff :( 

(I am not talking about relying on `node_modules` dependencies)

**Comment by Kureev** [comment]
> But you are relying on the package.json that was made for dependency tests, not for plugin system. 
I'm relying for a test fixture. If you see a case for splitting this package.json up - we can do it.

> I prefer mocking package.json on per-suite basis with just few lines instead of having 5 package.json just to test out all cases :)
I'll really prefer 5 package.json for all the cases (we can mock require with just an object for our needs, that'll allow us to avoid having 5 package.json if that's the case)

> Another thing is that we have ~40LOC of package.json just to test plugin system, which may get complicated in the future because we won't be sure what things our test suite uses from that package.json and what elements are useless. I mean - we do not need scripts nor version nor name to test findPlugins - that's why I prefer passing just part of it as I suggested.

Yeah, I agree, we can just use an object for this case.

**Comment by grabbou** [comment]
It's all in my PR :)

**Comment by grabbou** [comment]
Can I merge it or is there anything left you want to do?

**Comment by Kureev** [comment]
Yes, you can
